README
IPTables-Parse version 1.5
===========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module has no external module dependencies outside of those that
are included in perl Core.

COPYRIGHT AND LICENCE

This module is distributed under the same license as perl itself.

Copyright (C) 2005-2015 by Michael Rash

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.5 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension IPTables::Parse.

1.6 ...
    - Updated to use the '-w' argument on the iptables command line (if it is
      supported). This acquires an exclusive lock...

1.5 Mon Sep 07 20:18:16 2015
    - Bug fix to support additional characters in iptables chain names such as
      dashes and special characters. Stuart Schneider reported this bug and
      submitted a patch. Closes github issue #1.
    - Added 'rule_num' hash key to chain_rules() results (suggested by Stuart
      Schneider).
    - Updated _iptout and _ipterr tmp file paths to use the current PID as an
      extension by default if these files are not set when instantiating an
      IPTables::Parse object.  Also added a DESTROY block to clean these tmp
      files up. Patch submitted by Stuart Schneider.

1.4 Sun Mar 01 15:30:29 2015
    - Bump version to 1.4 so that CPAN recognizes it as a new version.

1.3.1 Sat Feb 28 07:15:55 2015
    - Fix version number in META.json and META.yml files.

1.3 Thu Feb 19 20:09:55 2015
    - Added support for systems with 'firewalld' via the 'firewall-cmd'
      command. Such systems include Fedora 21 for example.
    - Added list_table_chains() to return an array reference of all chains
      within the specified table. This feature was submitted as a patch by
      Fabien Mazieres
    - Added 'use_ipv6' and 'ipt_rules_file' keys to object constructor. These
      force ip6tables usage and specify a path to a file where
      'iptables -nL -v' output is included.
    - Updated perldoc documentation.

1.2 Sun Mar 04 21:21:11 2012
    - Major update to add a new 'parse_keys' hash to IPTables::Parse objects so
      that other modules can easily see what portions of iptables rules can be
      parsed into the data returned by chain_rules().
    - Added 'mac_source' into extended hash for parsed iptables rules.
    - Added support for the iptables 'length' match.

1.1 Fri Mar 02 22:31:12 2012
    - Minor update to print the iptables binary name in 'croak' error
      conditions.  The binary name is either 'iptables' or 'ip6tables'.
    - Minor perldoc updates to render links better (two spaces at the beginning
      of lines).

1.0 Tue Feb 28 21:45:19 2012
    - Added META.{yml,json} files to fix this bug:
        https://rt.cpan.org/Ticket/Display.html?id=75366
    - Added the ability to specify 'ip6tables' when instantiating an
      IPTables::ChainMgr object via 'new'.
    - Updated license to the Artistic license.

0.9 Sun Feb 26 21:01:45 2012
    - Applied slightly modified patch from SSIMON to properly pick up usage of
      state tracking in rule extended information as shown in this bug:
            https://rt.cpan.org/Ticket/Display.html?id=67372#txn-925687
      Rule 'extended' hash now includes the 'state' or 'ctstate' key
      depending on which iptables state tracking module is being used (if
      any).

0.8 Sun Feb 26 14:03:24 2012
    - Major update to support ip6tables policies.
    - Added test suite script t/basic_tests.pl to exercise major functions for
      both iptables and ip6tables.
    - Bugfix for default_log() and default_drop() functions to ensure that
      a proper return value is given in addition to the return of a results
      hash.
    - Migrated to git for source control:
            http://www.cipherdyne.org/cgi-bin/gitweb.cgi?p=IPTables-Parse.git;a=summary
            https://github.com/mrash/IPTables-Parse

0.7 Fri Oct 17 11:55:01 2008
    - Completely re-worked the manner in which iptables commands are executed
      so that they are sent through a single function with various options
      (described below) for controlling execution.
    - Added the ability to control iptables execution model.  The default is to
      use waitpid(), but other options are to use system() or popen().
    - Added the ability to introduce a configurable time delay between each
      iptables command.
    - Added the ability to use a function reference for the SIGCHLD signal
      handler.
    - Added the ability to configure the number of seconds used as the alarm
      timeout for iptables command execution in the waitpid() execution model.

0.6 Mon May 19 10:15:01 2008
    - Added perldoc documentation for 0.6 release.

0.3  12/18/2005
    - Added test for ULOG target
    - Update to allow -v iptables output (which may be supplied in a file).

0.01  Sat Feb  5 15:18:37 2005
    - original version; created by h2xs 1.23 with options
        -A -X -b 5.6.0 -n IPTables::Parse


snort_opts.pl
#!/usr/bin/perl -w
#
########################################################################
#
# File: snort_opts.pl
#
# Purpose: To parse snort rules and display a listing of snort fields
#          along with how many snort rules in which each field is
#          found.
#
########################################################################
#

use Getopt::Long;
use strict;

my $sig_file = '';
my $help = 0;

die "[*] Use --help for usage information.\n" unless (GetOptions(
    'file=s'    => \$sig_file,
    'help'      => \$help
));
&usage() if $help;

my %options = (
    'flow'         => 0,
    'flowbits'     => 0,
    'msg'          => 0,
    'logto'        => 0,
    'ttl'          => 0,
    'tos'          => 0,
    'id'           => 0,
    'ipopts'       => 0,
    'fragbits'     => 0,
    'dsize'        => 0,
    'flags'        => 0,
    'seq'          => 0,
    'ack'          => 0,
    'itype'        => 0,
    'icode'        => 0,
    'icmp_id'      => 0,
    'icmp_seq'     => 0,
    'content'      => 0,
    'uricontent'   => 0,
    'content-list' => 0,
    'offset'       => 0,
    'depth'        => 0,
    'nocase'       => 0,
    'file_data'    => 0,
    'rawbytes'     => 0,
    'session'      => 0,
    'rpc'          => 0,
    'resp'         => 0,
    'react'        => 0,
    'reference'    => 0,
    'sid'          => 0,
    'rev'          => 0,
    'classtype'    => 0,
    'priority'     => 0,
    'tag'          => 0,
    'ip_proto'     => 0,
    'sameip'       => 0,
    'asn1'         => 0,
    'stateless'    => 0,
    'regex'        => 0,
    'window'       => 0,
    'isdataat'     => 0,
    'distance'     => 0,
    'within'       => 0,
    'byte_jump'    => 0,
    'byte_test'    => 0,
    'byte_extract' => 0,
    'pcre'         => 0,
    'ftpbounce'    => 0,
    'base64_data'  => 0,
    'base64_decode' => 0,
    'http_header'  => 0,
    'http_cookie'  => 0,
    'http_uri'     => 0,
    'http_raw_uri' => 0,
    'urilen'       => 0,
    'http_method'  => 0,
    'http_stat_code' => 0,
    'http_stat_msg' => 0,
    'http_client_body' => 0,
    'fast_pattern' => 0,
    'metadata'     => 0,
    'threshold'    => 0,
    'detection_filter' => 0,
);

my %unrecognized = ();
my $total_rules = 0;
my $dir = 'deps/snort_rules';

my @rfiles = ();
if ($sig_file) {
    push @rfiles, $sig_file;
} else {
    opendir D, $dir or die "[*] Could not open $dir: $!";
    @rfiles = readdir D;
    closedir D;
}

print "[+] Calculating snort rule keyword percentages:\n";
for my $rfile (@rfiles) {

    my @lines = ();

    if ($sig_file) {
        open R, "< $sig_file" or die $!;
        @lines = <R>;
        close R;
    } else {
        next unless $rfile =~ /\.rules/;
        open R, "< $dir/$rfile" or die $!;
        @lines = <R>;
        close R;
    }

    for my $line (@lines) {
        chomp $line;
        next unless $line =~ /\S/;
        next if $line =~ /^#/;
        $total_rules++;
        if ($line =~ /^\s*alert/) {
            for my $opt (keys %options) {
                if ($line =~ /\s$opt[:;]/) {
                    $options{$opt}++;
                } elsif ($line =~ /\($opt[:;]/) {
                    $options{$opt}++;
                } elsif ($line =~ /;$opt[:;]/) {
                    $options{$opt}++;
                }
            }
            while ($line =~ m/[\s;](\w+)[:;]/g) {
                next if $1 =~ /^\d+$/;
                unless (defined $options{$1}) {
                    $unrecognized{$1}++;
                }
            }
        }
    }
}

my $max_opt_len = 0;
for my $opt (keys %options) {
    $max_opt_len = length($opt) if length($opt) > $max_opt_len;
}

for my $opt (sort {$options{$b} <=> $options{$a}} keys %options) {
    printf("%${max_opt_len}s %13s", $opt, "$options{$opt}/$total_rules  ");
    print sprintf("%.1f", $options{$opt} / $total_rules * 100) . "%\n";
}

print "\n[-] Potentially unrecognized options:\n";
for my $opt (keys %unrecognized) {
    print "    $opt\n";
}

exit 0;

sub usage() {
    print <<_USAGE_;
Usage: $0 [-f <sig file>] [-h]
_USAGE_
    exit 0;
}


snortspoof.pl
#!/usr/bin/perl -w
#
###############################################################################
#
# File: snortspoof.pl
#
# Purpose: To parse rules from the Snort rule set and spoof them at a target
#          IP from arbitrary source addresses.  This is similar to the
#          technique employed by the Stick and Snot projects.  Snortspoof.pl
#          is distributed with the fwsnort project
#          (http://www.cipherdyne.org/fwsnort/).
#
# Author: Michael Rash <mbr@cipherdyne.org>
#
# Copyright (C) 2003-2007 Michael Rash (mbr@cipherdyne.org)
#
# License (GNU Public License):
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301
#    USA
#
###############################################################################
#

require Net::RawIP;
use strict;

my $file       = $ARGV[0] || '';
my $spoof_addr = $ARGV[1] || '';
my $dst_addr   = $ARGV[2] || '';

die "$0 <rules file> <spoof IP> <dst IP>"
    unless $file and $spoof_addr and $dst_addr;

# alert udp $EXTERNAL_NET any -> $HOME_NET 635 (msg:"EXPLOIT x86 Linux mountd
# overflow"; content:"^|B0 02 89 06 FE C8 89|F|04 B0 06 89|F";
# reference:bugtraq,121; reference:cve,1999-0002; classtype:attempted-admin;
# sid:315; rev:6;)
my $sig_sent = 0;
open F, "< $file" or die "[*] Could not open $file: $!";
SIG: while (<F>) {
    my $content = '';
    my $conv_content = '';
    my $hex_mode = 0;
    my $proto = '';
    my $spt = 10000;
    my $dpt = 10000;

    ### make sure it is an inbound sig
    if (/^\s*alert\s+(tcp|udp)\s+\S+\s+(\S+)\s+\S+
            \s+(\$HOME_NET|any)\s+(\S+)\s/x) {
        $proto = $1;
        my $spt_tmp = $2;
        my $dpt_tmp = $4;

        ### can't handle multiple content fields yet
        next SIG if /content:.*\s*content\:/;

        $content = $1 if /\s*content\:\"(.*?)\"\;/;
        next SIG unless $content;

        if ($spt_tmp =~ /(\d+)/) {
            $spt = $1;
        } elsif ($spt_tmp ne 'any') {
            next SIG;
        }
        if ($dpt_tmp =~ /(\d+)/) {
            $dpt = $1;
        } elsif ($dpt_tmp ne 'any') {
            next SIG;
        }

        my @chars = split //, $content;
        for (my $i=0; $i<=$#chars; $i++) {
            if ($chars[$i] eq '|') {
                $hex_mode == 0 ? ($hex_mode = 1) : ($hex_mode = 0);
                next;
            }
            if ($hex_mode) {
                next if $chars[$i] eq ' ';
                $conv_content .= sprintf("%c",
                        hex($chars[$i] . $chars[$i+1]));
                $i++;
            } else {
                $conv_content .= $chars[$i];
            }
        }
        my $rawpkt = '';
        if ($proto eq 'tcp') {
            $rawpkt = new Net::RawIP({'ip' => {
                saddr => $spoof_addr, daddr => $dst_addr},
                'tcp' => { source => $spt, dest => $dpt, 'ack' => 1,
                data => $conv_content}})
                    or die "[*] Could not get Net::RawIP object: $!";
        } else {
            $rawpkt = new Net::RawIP({'ip' => {
                saddr => $spoof_addr, daddr => $dst_addr},
                'udp' => { source => $spt, dest => $dpt,
                data => $conv_content}})
                    or die "[*] Could not get Net::RawIP object: $!";
        }
        $rawpkt->send();
        $sig_sent++;
    }
}
print "[+] $file, $sig_sent attacks sent.\n";
close F;
exit 0;


install.pl
#!/usr/bin/perl -w
#
#######################################################################
#
# File: install.pl
#
# Purpose: This is the installation script for fwsnort.
#
# Copyright (C) 2003-2011 Michael Rash (mbr@cipherdyne.org)
#
# License (GNU Public License):
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301
#    USA
#
# TODO:
#   - Write the uninstall() routine.
#
#######################################################################
#

use Cwd;
use IO::Socket;
use File::Copy;
use File::Path;
use Getopt::Long;
use strict;

#========================= config ========================
my $fwsnort_conf_file = 'fwsnort.conf';

my $sbin_dir     = '/usr/sbin';
my $install_root = '/';

my $update_website = 'www.emergingthreats.net';

### system binaries
my $perlCmd = '/usr/bin/perl';
my $makeCmd = '/usr/bin/make';
my $wgetCmd = '/usr/bin/wget';
my $gzipCmd = '/bin/gzip';
my $tarCmd  = '/bin/tar';
#======================= end config ======================

my %config = ();

my @cmd_search_paths = qw(
    /bin
    /sbin
    /usr/bin
    /usr/sbin
    /usr/local/bin
    /usr/local/sbin
);

my %exclude_cmds = (
    'ip6tables' => ''
);

### map perl modules to versions
my %required_perl_modules = (
    'NetAddr::IP' => {
        'force-install' => 0,
        'mod-dir' => 'NetAddr-IP'
    },
    'IPTables::Parse' => {
        'force-install' => 1,
        'mod-dir' => 'IPTables-Parse'
    }
);

### rules update link
my $rules_url = 'http://rules.emergingthreats.net/open/snort-2.9.0/emerging-all.rules';

### establish some defaults
my $uninstall = 0;
my $skip_module_install   = 0;
my $cmdline_force_install = 0;
my $install_test_dir = 0;
my $force_mod_re = '';
my $exclude_mod_re = '';
my $deps_dir = 'deps';
my $is_root = 0;
my $help = 0;
my $locale = 'C';  ### default LC_ALL env variable
my $no_locale = 0;

my $src_dir = getcwd() or die "[*] Could not get current working directory.";

my %cmds = (
    'perl' => $perlCmd,
    'make' => $makeCmd,
    'gzip' => $gzipCmd,
    'wget' => $wgetCmd,
    'tar'  => $tarCmd
);

### make Getopts case sensitive
Getopt::Long::Configure('no_ignore_case');

&usage(1) unless (GetOptions(
    'force-mod-install' => \$cmdline_force_install,  ### force install of all modules
    'Force-mod-regex=s' => \$force_mod_re, ### force specific mod install with regex
    'Exclude-mod-regex=s' => \$exclude_mod_re, ### exclude a particular perl module
    'Skip-mod-install'  => \$skip_module_install,
    'rules-url=s' => \$rules_url,
    'uninstall' => \$uninstall, ### uninstall fwsnort
    'install-test-dir'  => \$install_test_dir,
    'LC_ALL=s'  => \$locale,
    'no-LC_ALL' => \$no_locale,
    'help'      => \$help
));

&usage(0) if $help;

### set LC_ALL env variable
$ENV{'LC_ALL'} = $locale unless $no_locale;

### make a copy of the original fwsnort.conf file and restore at the end
copy $fwsnort_conf_file, "${fwsnort_conf_file}.orig" or die "[*] Could not ",
    "copy $fwsnort_conf_file -> $fwsnort_conf_file.orig";

if ($install_test_dir) {
    $install_root = getcwd() . '/test/fwsnort-install';
}

&import_config();

$force_mod_re = qr|$force_mod_re| if $force_mod_re;
$exclude_mod_re = qr|$exclude_mod_re| if $exclude_mod_re;

### see if the deps/ directory exists, and if not then we are installing
### from the -nodeps sources so don't install any perl modules
$skip_module_install = 1 unless -d $deps_dir;

### make sure the system binaries are where we think they are.
&check_commands();

### check to make sure we are running as root
 die "You need to be root (or equivalent UID 0",
    " account) to install/uninstall fwsnort!\n" unless &is_root();

if ($uninstall) {
    &uninstall();
} else {
    &install()
}

### restore the original fwsnort.conf file (this is just the local one in the
### sources directory).
if (-e "${fwsnort_conf_file}.orig") {
    unlink $fwsnort_conf_file if -e $fwsnort_conf_file;
    move "${fwsnort_conf_file}.orig", $fwsnort_conf_file;
}

exit 0;
#===================== end main ===================

sub install() {
    die "[*] You must run install.pl from the fwsnort " .
        "sources directory." unless -e 'fwsnort' and -e 'fwsnort.conf';

    unless (-d $config{'CONF_DIR'}) {
        &full_mkdir($config{'CONF_DIR'}, 0500);
    }
    unless (-d $config{'RULES_DIR'}) {
        &full_mkdir($config{'RULES_DIR'}, 0500);
    }

    ### install perl modules
    unless ($skip_module_install) {
        for my $module (keys %required_perl_modules) {
            &install_perl_module($module);
        }
    }
    chdir $src_dir or die "[*] Could not chdir $src_dir: $!";

    my $local_rules_dir = 'deps/snort_rules';
    if (-d 'deps' and -d $local_rules_dir
            and &query_get_emerging_threats_sigs()) {
        chdir $local_rules_dir or die "[*] Could not chdir $local_rules_dir";
        if (-e 'emerging-all.rules') {
            move 'emerging-all.rules', 'emerging-all.rules.tmp'
                or die "[*] Could not move emerging-all.rules -> ",
                "emerging-all.rules.tmp";
        }
        system "$cmds{'wget'} $rules_url";
        if (-e 'emerging-all.rules') {  ### successful download
            unlink 'emerging-all.rules.tmp';
        } else {
            print "[-] Could not download emerging-all.rules file.\n";
            if (-e 'emerging-all.rules.tmp') {
                ### move the original back
                move 'emerging-all.rules', 'emerging-all.rules.tmp'
                    or die "[*] Could not move emerging-all.rules -> ",
                    "emerging-all.rules.tmp";
            }
        }
        chdir '../..';
    }

    if (-d 'deps' and -d $local_rules_dir) {
        opendir D, $local_rules_dir or die "[*] Could not open ",
            "the $local_rules_dir directory: $!";
        my @rfiles = readdir D;
        closedir D;

        print "[+] Copying all rules files to $config{'RULES_DIR'}\n";
        for my $rfile (@rfiles) {
            next unless $rfile =~ /\.rules$/;
            print "[+] Installing $rfile\n";
            copy "$local_rules_dir/${rfile}", "$config{'RULES_DIR'}/${rfile}"
                or die "[*] Could not copy $local_rules_dir/${rfile} ",
                    "-> $config{'RULES_DIR'}/${rfile}";
        }
    }

    print "\n";

    ### install the fwsnort.8 man page
    &install_manpage();

    my $preserve_rv = 0;
    if (-e "$config{'CONF_DIR'}/fwsnort.conf") {
        $preserve_rv = &query_preserve_config();
    }

    if ($preserve_rv) {
        &preserve_config();
    } else {
        print "[+] Copying fwsnort.conf -> $config{'CONF_DIR'}/fwsnort.conf\n";
        copy 'fwsnort.conf', "$config{'CONF_DIR'}/fwsnort.conf";
        chmod 0600, "$config{'CONF_DIR'}/fwsnort.conf";
    }

    print "[+] Copying fwsnort -> ${sbin_dir}/fwsnort\n";
    copy 'fwsnort', "${sbin_dir}/fwsnort";
    chmod 0500, "${sbin_dir}/fwsnort";

    print "\n========================================================\n",
        "\n[+] fwsnort will generate an iptables script located at:\n",
        "    /var/lib/fwsnort.sh when executed.\n",
        "\n[+] fwsnort has been successfully installed!\n\n";

    return;
}

sub install_perl_module() {
    my $mod_name = shift;

    chdir $src_dir or die "[*] Could not chdir $src_dir: $!";
    chdir $deps_dir or die "[*] Could not chdir($deps_dir): $!";

    die '[*] Missing force-install key in required_perl_modules hash.'
        unless defined $required_perl_modules{$mod_name}{'force-install'};
    die '[*] Missing mod-dir key in required_perl_modules hash.'
        unless defined $required_perl_modules{$mod_name}{'mod-dir'};

    if ($exclude_mod_re and $exclude_mod_re =~ /$mod_name/) {
        print "[+] Excluding installation of $mod_name module.\n";
        return;
    }

    my $version = '(NA)';

    my $mod_dir = $required_perl_modules{$mod_name}{'mod-dir'};

    if (-e "$mod_dir/VERSION") {
        open F, "< $mod_dir/VERSION" or
            die "[*] Could not open $mod_dir/VERSION: $!";
        $version = <F>;
        close F;
        chomp $version;
    } else {
        print "[-] Warning: VERSION file does not exist in $mod_dir\n";
    }

    my $install_module = 0;

    if ($required_perl_modules{$mod_name}{'force-install'}
            or $cmdline_force_install) {
        ### install regardless of whether the module may already be
        ### installed
        $install_module = 1;
    } elsif ($force_mod_re and $force_mod_re =~ /$mod_name/) {
        print "[+] Forcing installation of $mod_name module.\n";
        $install_module = 1;
    } else {
        if (has_perl_module($mod_name)) {
            print "[+] Module $mod_name is already installed in the ",
                "system perl tree, skipping.\n";
        } else {
            ### install the module in the /usr/lib/fwsnort directory because
            ### it is not already installed.
            $install_module = 1;
        }
    }

    if ($install_module) {
        unless (-d $config{'LIBS_DIR'}) {
            print "[+] Creating $config{'LIBS_DIR'}\n";
            &full_mkdir($config{'LIBS_DIR'}, 0755);
        }
        print "[+] Installing the $mod_name $version perl " .
            "module in $config{'LIBS_DIR'}/\n";
        my $mod_dir = $required_perl_modules{$mod_name}{'mod-dir'};
        chdir $mod_dir or die "[*] Could not chdir to ",
            "$mod_dir: $!";
        unless (-e 'Makefile.PL') {
            die "[*] Your $mod_name source directory appears to be incomplete!\n",
                "    Download the latest sources from ",
                "http://www.cipherdyne.org/\n";
        }
        system "$cmds{'make'} clean" if -e 'Makefile';
        system "$cmds{'perl'} Makefile.PL " .
            "PREFIX=$config{'LIBS_DIR'} LIB=$config{'LIBS_DIR'}";
        system $cmds{'make'};
#        system "$cmds{'make'} test";
        system "$cmds{'make'} install";

        print "\n\n";
    }
    chdir $src_dir or die "[*] Could not chdir $src_dir: $!";
    return;
}

sub has_perl_module() {
    my $module = shift;

    # 5.8.0 has a bug with require Foo::Bar alone in an eval, so an
    # extra statement is a workaround.
    my $file = "$module.pm";
    $file =~ s{::}{/}g;
    eval { require $file };

    return $@ ? 0 : 1;
}

sub uninstall() {
    ### FIXME
    return;
}

sub install_manpage() {
    my $manpage = 'fwsnort.8';
    ### remove old man page
    unlink "/usr/local/man/man8/${manpage}" if
        (-e "/usr/local/man/man8/${manpage}");

    ### default location to put the fwsnort man page, but check with
    ### /etc/man.config
    my $mpath = '/usr/share/man/man8';
    if (-e '/etc/man.config') {
        ### prefer to install $manpage in /usr/local/man/man8 if
        ### this directory is configured in /etc/man.config
        open M, '< /etc/man.config' or
            die "[*] Could not open /etc/man.config: $!";
        my @lines = <M>;
        close M;
        ### prefer the path "/usr/share/man"
        my $found = 0;
        for my $line (@lines) {
            chomp $line;
            if ($line =~ m|^MANPATH\s+/usr/share/man|) {
                $found = 1;
                last;
            }
        }
        ### try to find "/usr/local/man" if we didn't find /usr/share/man
        unless ($found) {
            for my $line (@lines) {
                chomp $line;
                if ($line =~ m|^MANPATH\s+/usr/local/man|) {
                    $mpath = '/usr/local/man/man8';
                    $found = 1;
                    last;
                }
            }
        }
        ### if we still have not found one of the above man paths,
        ### just select the first one out of /etc/man.config
        unless ($found) {
            for my $line (@lines) {
                chomp $line;
                if ($line =~ m|^MANPATH\s+(\S+)|) {
                    $mpath = $1;
                    last;
                }
            }
        }
    }
    &full_mkdir($mpath, 0755);
    my $mfile = "${mpath}/${manpage}";
    print "[+] Installing $manpage man page as $mfile\n";
    copy $manpage, $mfile or die "[*] Could not copy $manpage to " .
        "$mfile: $!";
    chmod 0644, $mfile;
    print "[+] Compressing manpage $mfile\n";
    ### remove the old one so gzip doesn't prompt us
    unlink "${mfile}.gz" if -e "${mfile}.gz";
    system "$gzipCmd $mfile";
    return;
}

sub query_get_emerging_threats_sigs() {
    return 0 if $install_test_dir;
    my $ans = '';
    print "[+] Would you like to download the latest Snort rules from \n",
        "    http://$update_website/?\n";
    while ($ans ne 'y' && $ans ne 'n') {
        print "    ([y]/n)?  ";
        $ans = <STDIN>;
        return 1 if $ans eq "\n";
        chomp $ans;
    }
    if ($ans eq 'y') {
        return 1;
    }
    return 0;
}

sub import_config() {
    open C, "< $fwsnort_conf_file"
        or die "[*] Could not open $fwsnort_conf_file: $!";
    while (<C>) {
        next if /^\s*#/;
        if (/^\s*(\S+)\s+(.*?)\;/) {
            my $varname = $1;
            my $val     = $2;
            if ($val =~ m|/.+| and $varname =~ /^\s*(\S+)Cmd$/) {
                ### found a command
                $cmds{$1} = $val;
            } else {
                $config{$varname} = $val;
            }
        }
    }
    close C;

    ### see if the install root is the same as the default in fwsnort.conf and
    ### update if not
    if ($install_root ne '/') {
        $install_root = getcwd() . "/$install_root"
            unless $install_root =~ m|^/|;
        $config{'INSTALL_ROOT'} = $install_root;
        $sbin_dir = $config{'INSTALL_ROOT'} . $sbin_dir;

        &put_var('INSTALL_ROOT', $install_root, $fwsnort_conf_file);
    }

    ### resolve internal vars within variable values
    &expand_vars();

    for my $dir ($install_root,
            $sbin_dir,
            $config{'LOG_DIR'},
            $config{'LIBS_DIR'},
            $config{'STATE_DIR'},
            $config{'QUEUE_RULES_DIR'},
            $config{'ARCHIVE_DIR'},
        ) {
        &full_mkdir($dir, 0755) unless -d $dir;
    }

    &required_vars();

    return;
}

sub expand_vars() {

    my $has_sub_var = 1;
    my $resolve_ctr = 0;

    while ($has_sub_var) {
        $resolve_ctr++;
        $has_sub_var = 0;
        if ($resolve_ctr >= 20) {
            die "[*] Exceeded maximum variable resolution counter.";
        }
        for my $hr (\%config, \%cmds) {
            for my $var (keys %$hr) {
                my $val = $hr->{$var};
                if ($val =~ m|\$(\w+)|) {
                    my $sub_var = $1;
                    die "[*] sub-ver $sub_var not allowed within same ",
                        "variable $var" if $sub_var eq $var;
                    if (defined $config{$sub_var}) {
                        if ($sub_var eq 'INSTALL_ROOT' and $config{$sub_var} eq '/') {
                            $val =~ s|\$$sub_var||;
                        } else {
                            $val =~ s|\$$sub_var|$config{$sub_var}|;
                        }
                        $hr->{$var} = $val;
                    } else {
                        die "[*] sub-var \"$sub_var\" not defined in ",
                            "config for var: $var."
                    }
                    $has_sub_var = 1;
                }
            }
        }
    }
    return;
}

sub put_var() {
    my ($var, $value, $file) = @_;

    open RF, "< $file" or die "[*] Could not open $file: $!";
    my @lines = <RF>;
    close RF;
    open F, "> $file" or die "[*] Could not open $file: $!";
    for my $line (@lines) {
        if ($line =~ /^\s*$var\s+.*;/) {
            printf F "%-24s%s;\n", $var, $value;
        } else {
            print F $line;
        }
    }
    close F;
    return;
}

sub required_vars() {
    my @required_vars = qw(
        CONF_DIR RULES_DIR ARCHIVE_DIR QUEUE_RULES_DIR LOG_DIR LIBS_DIR
        CONF_FILE FWSNORT_SCRIPT LOG_FILE
    );
    for my $var (@required_vars) {
        die "[*] Variable $var not defined in $fwsnort_conf_file. Exiting.\n"
            unless defined $config{$var};
    }
    return;
}

### check paths to commands and attempt to correct if any are wrong.
sub check_commands() {

    CMD: for my $cmd (keys %cmds) {
        next CMD if defined $exclude_cmds{$cmd};
        unless (-x $cmds{$cmd}) {
            my $found = 0;
            PATH: for my $dir (@cmd_search_paths) {
                if (-x "${dir}/${cmd}") {
                    $cmds{$cmd} = "${dir}/${cmd}";
                    $found = 1;
                    last PATH;
                }
            }
            unless ($found) {
                die "\n[*] Could not find $cmd anywhere!!!  ",
                    "Please edit the config section to include the path to ",
                    "$cmd.";
            }
        }
        unless (-x $cmds{$cmd}) {
            return unless &is_root();
            die "\n[*] $cmd is located at ",
                "$cmds{$cmd} but is not executable by uid: $<";
        }
    }
    return;
}

sub is_root() {
    if ($< == 0 and $> == 0) {
        $is_root = 1;
    }
    return $is_root;
}

sub query_preserve_config() {
    return 0 if $install_test_dir;
    my $ans = '';
    while ($ans ne 'y' && $ans ne 'n') {
        print "[+] Would you like to preserve the config from the\n",
            '    existing fwsnort installation ([y]/n)?  ';
        $ans = <STDIN>;
        return 1 if $ans eq "\n";
        chomp $ans;
    }
    if ($ans eq 'y') {
        return 1;
    }
    return 0;
}

sub preserve_config() {
    my $file = 'fwsnort.conf';
    open C, "< $file" or die "[*] Could not open $file: $!";
    my @new_lines = <C>;
    close C;

    open CO, "< $config{'CONF_DIR'}/$file" or die "[*] Could not open ",
        "$config{'CONF_DIR'}/$file: $!";
    my @orig_lines = <CO>;
    close CO;

    print "[+] Preserving existing config: $config{'CONF_DIR'}/$file\n";
    ### write to a tmp file and then move.
    my $printed_intf_warning = 0;
    open CONF, "> $config{'CONF_DIR'}/${file}.new" or die "[*] Could not open ",
        "$config{'CONF_DIR'}/${file}.new: $!";
    for my $new_line (@new_lines) {
        if ($new_line =~ /^\s*#/) {
            print CONF $new_line;
        } elsif ($new_line =~ /^\s*(\S+)/) {
            my $var = $1;
            my $found = 0;
            for my $orig_line (@orig_lines) {
                if ($orig_line =~ /^\s*\S+INTF\s/) {
                    ### interfaces are no longer used!
                    unless ($printed_intf_warning) {
                        print "    NOTE: Interfaces are no longer used as of the ",
                        "0.8.0 release;\n    removing $var\n";
                        $printed_intf_warning = 1;
                    }
                }
                if ($orig_line =~ /^\s*$var\s/
                        and $orig_line !~ /INTF/) {
                    print CONF $orig_line;
                    $found = 1;
                    last;
                }
            }
            unless ($found) {
                print CONF $new_line;
            }
        } else {
            print CONF $new_line;
        }
    }
    close CONF;
    move "$config{'CONF_DIR'}/${file}.new", "$config{'CONF_DIR'}/$file";
    return;
}

sub full_mkdir() {
    my ($dir, $perms) = @_;

    my @dirs = split /\//, $dir;
    my $path = $dirs[0];
    shift @dirs;
    for my $d (@dirs) {
        next unless $d and $d =~ /\S/;
        $path .= "/$d";
        unless (-d $path) {
            printf "[+] mkdir $path, %o\n", $perms;
            mkdir $path, $perms or die "[*] Could not mkdir($path): $!";
        }
    }
    return;
}

sub usage() {
    my $exit = shift;
    print <<_HELP_;
install.pl: [options]
    -f, --force-mod-install        - Install all perl modules regardless
                                     of whether they already installed on
                                     the system.
    -F, --Force-mod-regex <regex>  - Install perl module that matches a
                                     specific regular expression.
    -E, --Exclude-mod-regex <re>   - Exclude a perl module that matches this
                                     regular expression.
    -S, --Skip-mod-install         - Skip installation of modules.
    -r, --rules-url <url>          - Specify the URL to use for updating the
                                     Emerging Threats rule set - the default is:
                                     $rules_url
    --install-test-dir             - Install fwsnort in test/fwsnort-install
                                     for test suite.
    -u, --uninstall   - uninstall fwsnort
    -h, --help        - print help and exit
_HELP_
    exit $exit;
}
